File: minesweeper.c
#include <stdio.h>
#include <assert.h>
#include "minesweeper.h"

//////////////////////////////////////
// CONSTANTS:
const char UNREVEALED = ' ' ;
const char REVEALED[9] = "012345678";
const char FLAG = 'F';
const char MINE = '*';
//////////////////////////////////////


// within_bounds(b, x, y) checks if location (x, y) is within the boundaries of *b
// requires: *b is a valid board
// time: O(1)
static bool within_bounds(struct board *b, int x, int y) {
	return x > 0 && x <= b->width && y > 0 && y <= b->height;
}

// is_mine(b, x, y) returns true if there is a mine at location (x, y) on *b, 
// else returns false
// note: returns false for invalid locations 
// requires: *b is a valid board
// time: O(n) where n is the number of mines in *b
static bool is_mine(struct board *b, int x, int y) {
	assert(b);
	if (within_bounds(b, x, y)) {
		for (int i = 0; i < b->num_mines; i++) {
			if (b->mines[i].x == x && b->mines[i].y == y) {
				return true;
			}
		}
	}
	return false;
}

bool flag(struct board *b, int x, int y) {
	assert(b);
	if (!within_bounds(b, x, y)) {
		return false;
	}
	char *tile = &(b->grid)[(y - 1) * b->width + x - 1];
	if (*tile == UNREVEALED) {
		*tile = FLAG;
		return true;
	} else if (*tile == FLAG) {
		*tile = UNREVEALED;
		return true;
	}
	return false;
}

// count_mines(b, x, y) returns the number of mines adjacent to (x, y) on *b
// requires: *b is a valid board
// time: O(n) where n is the number of mines in *b
static int count_mines(struct board *b, int x, int y) {
	assert(b);
	int count = 0;
	for (int x_offset = -1; x_offset <= 1; x_offset++) {
		for (int y_offset = -1; y_offset <= 1; y_offset++) {
			if (is_mine(b, x + x_offset, y + y_offset)) {
				count++;
	  		}
		}
	}
	return count;
}

bool reveal(struct board *b, int x, int y) {
	assert(b);
	if (!within_bounds(b, x, y)) {
		return false;
	}
	char *tile = &(b->grid)[(y - 1) * b->width + x - 1];
	if (*tile != UNREVEALED || *tile == FLAG) {
		return false;
	} else if (is_mine(b, x, y)) {
		*tile = MINE;
		return true;
	}
	int z = count_mines(b, x, y);
	*tile = REVEALED[z];
	if (z == 0) {
		for (int x_offset = -1; x_offset <= 1; x_offset++) {
	  		for (int y_offset = -1; y_offset <= 1; y_offset++) {
				if (within_bounds(b, x + x_offset, y + y_offset)) {
					reveal(b, x + x_offset, y + y_offset);
				}
	  		}
		}
	}
	return true;
}

bool game_won(const struct board *b) {
	assert(b);
	int num_revealed = 0;
	for (int x = 1; x <= b->width; x++) {
		for (int y = 1; y <= b->height; y++) {
			char tile = (b->grid)[(y - 1) * b->width + x - 1];
			if (tile == MINE) {
				return false;
			}
			if (tile != UNREVEALED && tile != FLAG) {
				num_revealed++;
			}
		}
	}
	if (num_revealed == (b->width) * (b->height) - (b->num_mines)) {
		return true;
	} else {
		return false;
	}
}

bool game_lost(const struct board *b) {
	assert(b);
	for (int i = 0; i < b->num_mines; i++) {
		int x = b->mines[i].x;
		int y = b->mines[i].y;
		char tile = (b->grid)[(y - 1) * b->width + x - 1];
		if (tile == MINE) {
	  		return true;
		}
	}
	return false;
}

File: minepy/minesweeper.py
import random

UNREVEALED = " "
FLAG = "F"
MINE = "X"

class Minesweeper:
    """Minesweeper game base class."""

    def __init__(self, width, height, mine_count):
        self.width = width
        self.height = height
        self.mine_count = mine_count

        self.revealed_count = 0
        self.mines_remaining = mine_count
        self.game_over = False

        self.board = []
        for y in range(self.height):
            row = []
            for x in range(self.width):
                row.append(UNREVEALED)

            self.board.append(row)

        self.mines = set()
        self.__generate_mines()


    def __generate_mines(self):
        """Randomly generate mines."""
        while len(self.mines) != self.mine_count:
            mine_x = random.randint(0, self.width - 1)
            mine_y = random.randint(0, self.height - 1)
            self.mines.add((mine_x, mine_y))


    def flag(self, x, y):
        """Flag/unflag square (x, y)."""
        if x in range(self.width) and y in range(self.height):
            if self.board[y][x] == UNREVEALED:
                self.board[y][x] = FLAG
                self.mines_remaining -= 1
            elif self.board[y][x] == FLAG:
                self.board[y][x] = UNREVEALED
                self.mines_remaining += 1


    def __neighbours(self, x, y):
        """Return a list of neighbours of square (x, y)."""
        return ((a, b) for a in (x - 1, x, x + 1) for b in (y - 1, y, y + 1))


    def __count_mines(self, x, y):
        """Return the number of mines adjacent to square (x, y)."""
        count = 0

        for (i, j) in self.__neighbours(x, y):
            if (i, j) in self.mines:
                count += 1

        return count


    def reveal(self, x, y):
        """Reveal square (x, y)."""
        if x in range(self.width) and y in range(self.height) and self.board[y][x] == UNREVEALED:
            if (x, y) in self.mines:
                self.game_over = True
                self.board[y][x] = MINE

                for (i, j) in self.mines:
                    if self.board[j][i] != FLAG:
                        self.board[j][i] = MINE
            else:
                self.board[y][x] = self.__count_mines(x, y)
                self.revealed_count += 1

                if self.board[y][x] == 0:
                    for (i, j) in self.__neighbours(x, y):
                        self.reveal(i, j)

                if self.game_won():
                    for (i, j) in self.mines:
                        self.board[j][i] = FLAG


    def game_won(self):
        """Return true if game is won."""
        return self.revealed_count == self.width * self.height - self.mine_count


    def game_lost(self):
        """Return true if game is lost."""
        return self.game_over
